lib/queryContextCache.js

```diff
@@ -0,0 +1,281 @@
+  Logger.getInstance().debug(
+    `checkCacheCapacity() called. treeSet size ${this.treeSet.size} cache capacity ${this.capacity}` );
+
+  // remove elements based on priority
+  while (this.treeSet.size > this.capacity) {
+    const qce = Array.from(this.treeSet).pop();
+    this.removeQCE(qce);
+  }
+  Logger.getInstance().debug(
+    `checkCacheCapacity() returns. treeSet size ${this.treeSet.size} cache capacity ${this.capacity}`,
+  );
+};
+
+/** Clear the cache. */
+QueryContextCache.prototype.clearCache = function () {
+  Logger.getInstance().debug('clearCache() called');
+  this.idMap.clear();
+  this.priorityMap.clear();
+  this.treeSet.clear();
+  Logger.getInstance().debug(`clearCache() returns. Number of entries in cache now ${this.treeSet.size}`,);
+};
+
+QueryContextCache.prototype.getElements = function () {
+  return this.treeSet;
+};
+
+/**
+ * @param data: the QueryContext Object serialized as a JSON format string
+ */
+QueryContextCache.prototype.deserializeQueryContext = function (data) {
+  const stringifyData = JSON.stringify(data);
+  Logger.getInstance().debug(`deserializeQueryContext() called: data from server: ${stringifyData}`);
+  if (!data || stringifyData ==='{}' || data.entries === null) {
+
+    this.clearCache();
+    Logger.getInstance().debug('deserializeQueryContext() returns');
+    this.logCacheEntries();
+    return;
+  }
+  try {
+    // Deserialize the entries. The first entry with priority is the main entry. An example JSON is:
+    // {
+    //   "entries": [
+    //    {
+    //     "id": 0,
+    //     "readtimestamp": 123456789,
+    //     "priority": 0,
+    //     "context": "base64 encoded context"
+    //    },
+    //     {
+    //       "id": 1,
+    //       "readtimestamp": 123456789,
+    //       "priority": 1,
+    //       "context": "base64 encoded context"
+    //     },
+    //     {
+    //       "id": 2,
+    //       "readtimestamp": 123456789,
+    //       "priority": 2,
+    //       "context": "base64 encoded context"
+    //     }
+    //   ]
+
+    const entries = data.entries;
+    if (entries !== null && Array.isArray(entries)) {
+      for (const entryNode of entries) {
+        const entry = this.deserializeQueryContextElement(entryNode);
+        if (entry != null) {
+          this.merge(entry);
+        } else {
+          Logger.getInstance().warn(
+            'deserializeQueryContextJson: deserializeQueryContextElement meets mismatch field type. Clear the QueryContextCache.');
+          this.clearCache();
+          return;
+        }
+      }
+     
+    }
+  } catch (e) {
+    Logger.getInstance().debug(`deserializeQueryContextJson: Exception = ${e.getMessage}`, );
+
+    // Not rethrowing. clear the cache as incomplete merge can lead to unexpected behavior.
+    this.clearCache();
+  }
+
+  // After merging all entries, truncate to capacity
+  this.checkCacheCapacity();
+
+  // Log existing cache entries
```

Comment:
do we need to comment directly what we have in the next line?

Reply:
I do not think so. I removed them.